Implement Color Detection

FireShredder24 opened the pull request:
Adds a Subsystem, the ColorSubsystem, that handles the conversion of raw RGB color data from the REVRobotics ColorSensorV3 into one of the four colors on the Control Panel.

terbos90803 committed:
wip 1/14

terbos90803 committed:
Add auto-detected color field

terbos90803 committed:
wip part 2

Jedralin committed:
Fix vendordep & prep for testing

I reinstalled the REVColorSensorV3 vendordep so it works again.  Removed all speed controllers so the Driver Station won't throw a watchdog error when Chairlie Jr. doesn't have a CAN bus.
Authored by John

Jedralin committed:
Fix vendordep & prep for testing

I reinstalled the REVColorSensorV3 vendordep so it works again.  Removed all speed controllers so the Driver Station won't throw a watchdog error when Chairlie Jr. doesn't have a CAN bus.

Jedralin committed:
Merge remote-tracking branch 'origin/colorTesting' into colorTesting

Jedralin committed:
Change println() output to HSV

Changes the testPeriodic() println() to output color data in the HSV color space instead of the RGB color space.  This is to get a more consistent color reading from the REVRobotics sensor that doesn't largely depend on range.

Jedralin committed:
Change println() output to HSV

Changes the testPeriodic() println() to output color data in the HSV color space instead of the RGB color space.  This is to get a more consistent color reading from the REVRobotics sensor that doesn't largely depend on range.
Authored by John

Jedralin committed:
Save test logs

Adds the json files I recorded from testing
Authored by John

Jedralin committed:
Save test logs

Adds the json files I recorded from testing

FireShredder24 committed:
Merge remote-tracking branch 'origin/colorTesting' into colorTesting

FireShredder24 committed:
Tune RGYC qualifying

Tuned the TestPeriodic() function to accurately detect each of the 4 colors on the Control Panel.
Authored by John

FireShredder24 committed:
Prepare to migrate color detection to a Subsystem

Prep for migration of colorsensing code to a separate Subsystem so color data can be gotten() by Commands such as the ControlPanelSubsystem

FireShredder24 committed:
Finish ColorSensing class

Finished declaring the ColorSensing class and its sole method, String getColor(), with verbose comments.

Jedralin committed:
Prep for merge

Authored by John

FireShredder24 committed:
Migrate to a Subsystem

FireShredder24 committed:
Migrate to a Subsystem

Fix lack of EOF newlines

FireShredder24 committed:
Merge remote-tracking branch 'origin/colorTesting' into colorTesting

Fix lack of EOF newline in DriveSubsystem

FireShredder24 committed:
Merge remote-tracking branch 'origin/colorTesting' into colorTesting

FireShredder24 committed:
Merge remote-tracking branch 'origin/colorTesting' into colorTesting

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+   * @param hsv The array that is used to contain the output of
+   *            java.awt.Color.RGBtoHSB()
+   */
+  public float[] hsv = new float[3];
```
class members should not be public.  Make this private.  Can also make final because the Java semantics of final only prevent replacing the instance of the array, they don't block altering the contents of the array.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+   */
+  public float[] hsv = new float[3];
+  // The ints used to store the raw ADC output of the ColorSensorV3
+  public int red, green, blue;
```
Same comment applies to all these members.  They should be private.

Also, these variables should be local to the method, not class members.  The only reason that hsv is a class member is to avoid reallocating the array.  The rest of these should just be locals.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+ */
+public class ColorSubsystem extends SubsystemBase {
+  /**
+   * @param hsv The array that is used to contain the output of
```
This is not a @param.  Remove that tag.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+  public int red, green, blue;
+  // The maximum allowed absolute distance from the constant to ensure an accurate
+  // reading
+  final float tolerance = 0.025f;
```
This is not marked with a scope, which means it's package private.  Unless you have a really good reason for exposing constants, they should be explicitly private.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+  final float tolerance = 0.025f;
+  // The constants that correspond to the average hue (H) of the Control Panel
+  // colors.
+  final float kYellow = 0.25f;
```
To simplify the matching code, it might be better to group these constants as some form of collection.  The best collection type for this is probably a Map.  It associates keys (the number) with values (the color name).

The color names should probably be an enum set defined globally because it'll need to be shared across the APIs between subsystems and commands.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+  // The absolute difference between the measured value and the constants
+  private float diff;
+  // Used to store the return message
+  String color;
```
Why are you storing a return value as a public member?  Neither of those things should be true.  Return values should be local to their method.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+  // Used to store the return message
+  String color;
+  // The REVRobotics ColorSensorV3
+  final ColorSensorV3 colorSensor = new ColorSensorV3(Port.kOnboard);
```
Private.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+    // fulfilled
+    color = "Error";
+    // Gets the raw ADC values
+    red = colorSensor.getRed();
```
Don't call getRed, getGreen, and getBlue separately.  You should call getRawColor and then pick the fields out of that.  It prevents calling the sensor multiple times.

obtw, did you notice in the docs that they provide a getColor method that returns a color enum?  I suspect many teams will start with that.  The big problem with using that directly is that there are a gazillion enums and I find it unlikely that it'd land on the ones we want reliably.  We're much better off using the HSB transform.

The docs also appear to be out of sync with the code samples.  In the samples, Rev provides a color Matching class that seems to do the same thing we're doing.  I suspect they're probably using it under the hood for their getColor method.  Again, the problem I have is they're probably doing the matching in RGB space, which loses out on the simplicity and robustness of doing it in HSB space.

FireShredder24 replied:
I did notice their getColor and its ridiculous enum and decided exactly that.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+    Color.RGBtoHSB(red, green, blue, hsv);
+    hue = hsv[0];
+    mindiff = 1f;
+    // Gets the difference between the measured value and the first constant
```
Obvious comments are not helpful.  This applies to many of the comments from here on.  Except in particularly obscure coding examples, the reader should be able to figure out the what pretty easily.  Make sure your comments explain the *why* of something and not the what.
Never be afraid to write a single big block comment explaining an algorithm at the top of that code instead of tagging each line.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+    }
+    // Gets the difference between the measured value and the third constant
+    diff = Math.abs(hue - kGreen);
+    // I don't really need to tell you what this does, do I?
```
Really?  Snide comments are worse than no comments.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -0,0 +1,97 @@
+    hue = hsv[0];
+    mindiff = 1f;
+    // Gets the difference between the measured value and the first constant
+    diff = Math.abs(hue - kYellow);
```
The DRY principle applies to this section of code.  I know we wrote it this way for expediency, but it really needs to be refactored into a loop.  This will be possible when the color constants are collected in a Map that can be looped on.

terbos90803 replied:
Did you need assistance with this refactor?

FireShredder24 replied:
Yes, and also with the Map collection.  The documentation is very unhelpful.

FireShredder24 committed:
Apply Terry's suggestions

Change the scope of class members
Move the RGB output ints to members of getColor()
Improve comments
Change getColor() from returning a String to returning an enum
Get RGB from ColorSensorV3.getRawColor() instead of ColorSensorV3.getRed(), getGreen(), getBlue().

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorOptions.java:
```diff
@@ -0,0 +1,5 @@
+package frc.robot.subsystems;
```
Probably not for the subsystem package.  Is there a more central place for constants?

FireShredder24 replied:
Yes, moved it to Utils.

terbos90803 commented on src/main/java/frc/robot/subsystems/ColorSubsystem.java:
```diff
@@ -2,91 +2,93 @@
+  private final float kGreen = 0.35f;
+  private final float kCyan = 0.50f;
   // Used to prevent repeated indexing of hsv[]
   private float hue;
```
These variables cannot be class members, they must be locals to the method.

terbos90803 reviewed (changes requested):
Still have some cleanup and refactor to do.

FireShredder24 committed:
Merge branch 'master' into colorTesting

FireShredder24 committed:
Increase efficiency of getColor logic

I refactored the constants and enums into a HashMap collection that can be iterated through, increasing the efficiency

terbos90803 committed:
Add unit testing for ColorSubsystem

FireShredder24 committed:
Centralize ColorOptions enum

Moved ColorOptions to frc.robot.utils, a better place for it

FireShredder24 committed:
Fix build error with ColorSubsystemTest

ColorOptions wasn't properly imported

jordan-powers committed:
Revert "Fix build error with ColorSubsystemTest"

This reverts commit 7a26d9343e3b611ea46a8e5a2509f93d679abb74.

Don't "fix" test cases that are failing. Fix the code that's failing.

jordan-powers committed:
Fix import in Unit test

FireShredder24 committed:
Merge branch 'master' into colorTesting

FireShredder24 committed:
Add Cyan, Yellow, and Green tests

FireShredder24 committed:
Merge remote-tracking branch 'origin/colorTesting' into colorTesting

terbos90803 commented on src/test/java/frc/robot/subsystems/ColorSubsystemTest.java:
```diff
@@ -0,0 +1,92 @@
+  @Test
+  public void testRed() {
+    RawColor rc = new RawColor(255, 153, 0, 0);
+    System.out.format("rc: Red:%d Green:%d Blue:%d\n", rc.red, rc.green, rc.blue);
```
unit tests should not have print statements.  I used these for getting the initial test working.  They should all come out.

terbos90803 commented on src/test/java/frc/robot/subsystems/ColorSubsystemTest.java:
```diff
@@ -0,0 +1,92 @@
+
+    assertEquals(ColorOptions.YELLOW, co);
+  }
+}
```
In addition to the checks for the 4 intended colors, there need to be a set of tests to verify that other colors are correctly rejected.

FireShredder24 replied:
I picked 4 colors that are similar in RGB values to each of the 4 colors on the CP to confirm that the detection code is able to reject them.  It does.

FireShredder24 committed:
Add unit tests for close-to-actual values

FireShredder24 committed:
Merge branch 'master' into colorTesting

terbos90803 reviewed (approved):
Looks good.
